### PR TITLE
[OpenMP] Add unit tests for nextgen plugins

### PR DESCRIPTION
**Pull request by EthanLuisMcDonough**
This patch add three GTest unit tests that test plugin read and write operations. Tests can be compiled with `ninja -C runtimes/runtimes-bins LibomptUnitTests`.

### CODE REVIEW

File: openmp/libomptarget/unittests/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+add_custom_target(LibomptUnitTests)
+set_target_properties(LibomptUnitTests PROPERTIES FOLDER "Libomptarget tests")
```
Comment (jhuber6) [inline openmp/libomptarget/unittests/CMakeLists.txt:2]:
```suggestion
set_target_properties(LibomptUnitTests PROPERTIES FOLDER "Tests/UnitTests")
```
This is a weird folder name.

File: openmp/libomptarget/unittests/Plugins/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,26 @@
+set(LIBOMPTARGET_BUILD_AMDGPU_PLUGIN TRUE CACHE BOOL
+  "Whether to build AMDGPU plugin")
+set(LIBOMPTARGET_BUILD_CUDA_PLUGIN TRUE CACHE BOOL
+  "Whether to build CUDA plugin")
+
+set(PLUGINS_TEST_COMMON omptarget OMPT omptarget.devicertl)
+set(PLUGINS_TEST_SOURCES NextgenPluginsTest.cpp)
+set(PLUGINS_TEST_INCLUDE ${LIBOMPTARGET_INCLUDE_DIR})
+
+if (LIBOMPTARGET_BUILD_AMDGPU_PLUGIN)
```
Comment (jhuber6) [inline openmp/libomptarget/unittests/Plugins/CMakeLists.txt:10]:
```suggestion
if(TARGET omptarget.rtl.amdgpu)
```
We should be able to just check the existence of the target so long as we add the unit tests after the main source.

File: openmp/libomptarget/unittests/Plugins/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,26 @@
+set(LIBOMPTARGET_BUILD_AMDGPU_PLUGIN TRUE CACHE BOOL
+  "Whether to build AMDGPU plugin")
+set(LIBOMPTARGET_BUILD_CUDA_PLUGIN TRUE CACHE BOOL
+  "Whether to build CUDA plugin")
+
```
Comment (jhuber6) [inline openmp/libomptarget/unittests/Plugins/CMakeLists.txt:5]:
```suggestion
```
Aren't these already declared somewhere else?

File: openmp/libomptarget/unittests/Plugins/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  libomptarget_say("Skipping AMDGPU plugin unit tests")
+endif()
+
+if (LIBOMPTARGET_BUILD_CUDA_PLUGIN)
+  libomptarget_say("Building plugin unit tests for Cuda")
+  add_libompt_unittest(PluginsTestCu ${PLUGINS_TEST_SOURCES})
+  add_dependencies(PluginsTestCu ${PLUGINS_TEST_COMMON} omptarget.rtl.cuda)
+  target_link_libraries(PluginsTestCu PRIVATE ${PLUGINS_TEST_COMMON} omptarget.rtl.cuda)
+  target_include_directories(PluginsTestCu PRIVATE ${PLUGINS_TEST_INCLUDE})
+endif()
```
Comment (jhuber6) [inline openmp/libomptarget/unittests/Plugins/CMakeLists.txt:26]: Realistically this should be a loop over all the targets, just have the plugin append some CMake list of enabled targets, then we can just iterate it here.

**Review by jhuber6 — COMMENTED** [review]
Thanks, I've been wanting the ability to test stuff without relying on OpenMP codegen for awhile.

File: openmp/libomptarget/unittests/Plugins/NextgenPluginsTest.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
+#include "Shared/PluginAPI.h"
```
Comment (jdoerfert) [inline openmp/libomptarget/unittests/Plugins/NextgenPluginsTest.cpp:1]: The licence header is missing.

File: openmp/libomptarget/unittests/Plugins/NextgenPluginsTest.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+#include "omptarget.h"
+#include "gtest/gtest.h"
+
+const int DEVICE_ID = 0, DEVICE_TWO = 1;
```
Comment (jdoerfert) [inline openmp/libomptarget/unittests/Plugins/NextgenPluginsTest.cpp:5]: In a follow up we should iterate over all devices, not hardcode 2.

**Review by jdoerfert — COMMENTED** [review]
Thanks, I think after Joseph's comments are addressed and the licence is in, we can merge this.

File: openmp/libomptarget/unittests/Plugins/NextgenPluginsTest.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  if (__tgt_rtl_init_plugin() == OFFLOAD_FAIL ||
+      __tgt_rtl_init_device(ID) == OFFLOAD_FAIL) {
+    return OFFLOAD_FAIL;
+  }
+  setup_map[ID] = true;
+  return OFFLOAD_SUCCESS;
+}
+
+// Test plugin initialization
+TEST(NextgenPluginsTest, PluginInit) {
```
[the remark this answers is not in the document]
Comment (EthanLuisMcDonough) [reply openmp/libomptarget/unittests/Plugins/NextgenPluginsTest.cpp:31]: The global variables serve to keep track of which devices have been initialized. Devices can't be re-initialized, so I think the initialized devices would have to either be tracked through static class members or exist outside the test fixture to persist across tests. I don't *think* there's a way to check if a device has been initialized through the functions defined in PluginAPI.h, but I definitely could be overlooking something.

**Review by jdoerfert — APPROVED** [review]
I'm fine with getting this in and adapting it in-tree.

@jplehr concerns?

File: openmp/libomptarget/CMakeLists.txt [hunk 8]
```diff
@@ -123,3 +123,9 @@ add_subdirectory(tools)
 
 # Add tests.
 add_subdirectory(test)
+
+# Add unit tests if GMock/GTest is present
+if (EXISTS ${LLVM_THIRD_PARTY_DIR}/unittest)
+  add_subdirectory(${LLVM_THIRD_PARTY_DIR}/unittest ${CMAKE_CURRENT_BINARY_DIR}/third-party/unittest)
+  add_subdirectory(unittests)
+endif()
```
[the remark this answers is not in the document]
Comment (jhuber6) [reply openmp/libomptarget/CMakeLists.txt:131]: Pretty sure that's a name defined in the same way, we could use the same name if needed.

We also need these to run w/ `check-openmp` or similar. Long term I'd like language tests to be all all in a different directory so we can test things like CUDA, HIP, SYCL, OpenACC, C, C++, Fortran, etc.

File: openmp/libomptarget/unittests/Plugins/NextgenPluginsTest.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (jplehr) [reply openmp/libomptarget/unittests/Plugins/NextgenPluginsTest.cpp:31]: That makes sense, thanks for explaining.

**Comment by jplehr** [comment]
> I'm fine with getting this in and adapting it in-tree.
> 
> @jplehr concerns?

No, I think that makes sense.

**Comment by vzakhari** [comment]
Hi @EthanLuisMcDonough, @jdoerfert, do you know what might be causing this buildbot failure https://lab.llvm.org/buildbot/#/builders/270/builds/5927/steps/6/logs/stdio?

**Comment by vvereschaka** [comment]
looks like these changes break the llvm configuration when `openmp` is a part of `LLVM_ENABLE_PROJECTS`, but not in `LLVM_ENABLE_RUNTIMES`. It tries to create new build targets with already existing names.
```
  add_library cannot create target "llvm_gtest" because another target with
  the same name already exists.  The existing target is a static library
  created in source directory
  ```
 and
 ```
  add_library cannot create target "llvm_gtest_main" because another target
  with the same name already exists.  The existing target is a static library
  created in source directory
 ```

**Comment by EthanLuisMcDonough** [comment]
I think #76141 should resolve this. I've my patch out locally, and it seems to fix the issue. I hope you all don't mind me immediately merging it; I didn't want to keep the buildbot failing.

**Comment by vvereschaka** [comment]
Thank you @EthanLuisMcDonough

**Pull request merged**